### PR TITLE
Create api endpoint for serving files

### LINKED ISSUE #58: Provide a way to the end-user to view the originally uploaded files in two buffers.
After getting a similarities heatmap, the user must be able to view and compare the content of pair room related uploaded files in the browser.

We need to make that available!!!
This should be done after https://github.com/smart-online-judge/basic-online-judge/issues/40

### PR DESCRIPTION
**Pull request by gevorgyana**
- Users of `link/` endpoint are now able to download files from
the bucket with a 1-hr valid download link.

closes #58 

Signed-off-by: Artyom Gevorgyan <artemhevorhian@gmail.com>

### CODE REVIEW

File: web-service/src/api/controllers/link_controller.go [hunk 1]
```diff
@@ -0,0 +1,46 @@
...
+
+	presignedURL := s3support.PrepareViewFileURL(id, fileName)
+	if presignedURL == nil {
+		errorLogger.Println(err)
+		msg := fmt.Sprintf("Unable to find a file with such name %s", fileName)
+		http.Error(w, msg, http.StatusInternalServerError)
+		return
+	}
+
+	w.Header().Add("Content-Type", "application/json")
```
Comment (gevorgyana) [inline web-service/src/api/controllers/link_controller.go:43]: should be Set()

File: web-service/src/api/controllers/link_controller.go [hunk 2]
```diff
@@ -0,0 +1,46 @@
...
+	s3support "web-service/src/s3support"
+
+	guuid "github.com/google/uuid"
+)
+
+/*
+   type JSONResponse struct {
+	name string
+        }
+*/
```
Comment (gevorgyana) [inline web-service/src/api/controllers/link_controller.go:17]: Don't need this

File: web-service/src/s3support/wrap_minio.go [hunk 3]
```diff
@@ -118,3 +123,44 @@ func LoadFileByUUID(id guuid.UUID, fileName string) io.Reader {
...
+			break
+		}
+	}
+	if ok == false {
+		return nil
+	}
+	reqParams := make(url.Values)
+	val := fmt.Sprintf("attachment; filename=\"%s\"", fileName)
+	reqParams.Set("response-content-disposition", val)
+	presignedURL, err := minioClient.PresignedGetObject(rootCtx, bucketName, path.Join(id.String(), fileName), time.Second*24*60*60, reqParams)
```
Comment (gevorgyana) [inline web-service/src/s3support/wrap_minio.go:160]: No need for this much. 1 hr is okay as the comment says. Forgot to update, probably.

**Review by Pilipets — CHANGES_REQUESTED** [review]

File: web-service/src/api/controllers/link_controller.go [hunk 4]
```diff
@@ -0,0 +1,46 @@
...
+	if presignedURL == nil {
+		errorLogger.Println(err)
+		msg := fmt.Sprintf("Unable to find a file with such name %s", fileName)
+		http.Error(w, msg, http.StatusInternalServerError)
+		return
+	}
+
+	w.Header().Add("Content-Type", "application/json")
+	jsonEncoded := fmt.Sprintf("{\"link\": \"%s\"}", presignedURL.String())
+	w.Write([]byte(jsonEncoded))
```
[the remark this answers is not in the document]
Comment (gevorgyana) [reply web-service/src/api/controllers/link_controller.go:45]: Write accepts bytes, not strings

File: web-service/src/api/controllers/link_controller.go [hunk 5]
```diff
@@ -0,0 +1,46 @@
...
+	guuid "github.com/google/uuid"
+)
+
+/*
+   type JSONResponse struct {
+	name string
+        }
+*/
+
+func ServeFileViewByUUID(w http.ResponseWriter, req *http.Request) {
```
[the remark this answers is not in the document]
Comment (gevorgyana) [reply web-service/src/api/controllers/link_controller.go:19]: Will decline any type of requests except GET.
GetFileLinkById will use it, sir.

File: web-service/src/s3support/wrap_minio.go [hunk 6]
```diff
@@ -118,3 +123,44 @@ func LoadFileByUUID(id guuid.UUID, fileName string) io.Reader {
...
+		}
+		nextFileName := strings.Split(object.Key, "/")
+		res = append(res, nextFileName[1])
+	}
+	return res
+}
+
+func PrepareViewFileURL(id guuid.UUID, fileName string) *url.URL {
+	fileNames := listFilesByUUID(id)
+	fmt.Println(fileNames)
```
[the remark this answers is not in the document]
Comment (gevorgyana) [reply web-service/src/s3support/wrap_minio.go:146]: ok

File: web-service/src/s3support/wrap_minio.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (gevorgyana) [reply web-service/src/s3support/wrap_minio.go:160]: Must? It is really a dummy variable. It has no meaning for us, agree? We control the expiration time using the database.

File: web-service/src/api/controllers/link_controller.go [hunk 7]
```diff
@@ -0,0 +1,46 @@
...
+	id, err := guuid.Parse(id_str)
+	if err != nil {
+		errorLogger.Println(err)
+		http.Error(w, "Please provide a valid UUID4", http.StatusUnprocessableEntity)
+		return
+	}
+
+	presignedURL := s3support.PrepareViewFileURL(id, fileName)
+	if presignedURL == nil {
+		errorLogger.Println(err)
```
[the remark this answers is not in the document]
Comment (gevorgyana) [reply web-service/src/api/controllers/link_controller.go:37]: will remove this.

[the remark this answers is not in the document]
Comment (gevorgyana) [reply web-service/src/api/controllers/link_controller.go:37]: will remove this

File: web-service/src/s3support/wrap_minio.go [hunk 8]
```diff
@@ -118,3 +123,44 @@ func LoadFileByUUID(id guuid.UUID, fileName string) io.Reader {
 		return v
 	}
 }
+
+func listFilesByUUID(id guuid.UUID) []string {
+	var res []string
+	objectCh := minioClient.ListObjects(rootCtx, bucketName, minio.ListObjectsOptions{
+		Prefix:    id.String(),
+		Recursive: true,
```
[the remark this answers is not in the document]
Comment (gevorgyana) [reply web-service/src/s3support/wrap_minio.go:131]: it is necessary for correctness.

File: web-service/src/s3support/wrap_minio.go (hunk 3 above)
[the remark this answers is not in the document]
Comment (Pilipets) [reply web-service/src/s3support/wrap_minio.go:160]: I would assign it to ten minutes then.

**Review by Pilipets — APPROVED** [review]

**Pull request merged**